File: LeetCode_Algorithm/790(Domino and Tromino Tiling)/790.py
# 

class Solution:
    def numTilings(self, n: int) -> int:
        # sol1 39ms 85% 16.1MB 97%
        dp = [0,1,2,5] + [0] * (n - 3)
        for i in range(4, n+1):
            dp[i] = (2 * dp[i-1]  + dp[i-3]) % 1000000007
        return dp[n]
        
print(Solution().numTilings(30))

File: LeetCode_Algorithm/790(Domino and Tromino Tiling)/790.js
var numTilings = function(n) {
  // sol1 54ms 75% 43.8MB 53%
  let dp = [0,1,2,5]
  for (let i = 4; i <= n; i++ ){
    dp.push(0)
    dp[i] = (2 * dp[i-1] + dp[i-3]) % 1000000007
  }  
  return dp[n]
};